common/src/main/java/org/apache/sedona/common/Functions.java

```diff
@@ -730,4 +731,149 @@ public static Geometry collectionExtract(Geometry geometry) {
+             * set of input points, the standard Weiszfeld method gets stuck with a
+             * divide-by-zero.
+             *
+             * To get ourselves out of the hole, we follow an alternate procedure to
+             * get the next iteration, as described in:
+             *
+             * Vardi, Y. and Zhang, C. (2011) "A modified Weiszfeld algorithm for the
+             * Fermat-Weber location problem."  Math. Program., Ser. A 90: 559-566.
+             * DOI 10.1007/s101070100222
+             *
+             * Available online at the time of this writing at
+             * http://www.stat.rutgers.edu/home/cunhui/papers/43.pdf
+             */
+            if (hit) {
+                double dx = 0;
+                double dy = 0;
+                double dz = 0;
+                for (int i = 0; i < points.length; i++) {
+                    double distance = distances[i];
+                    if (distance > DBL_EPSILON) {
+                        Coordinate coordinate = points[i];
+                        dx += (coordinate.x - curr.x) / distance;
+                        dy += (coordinate.y - curr.y) / distance;
+                        dz += (coordinate.z - curr.z) / distance;
+                    }
+                }
+                double dSqr = Math.sqrt(dx*dx + dy*dy + dz*dz);
+                /* Avoid division by zero if the intermediate point is the median */
+                if (dSqr > DBL_EPSILON) {
+                    double rInv = Math.max(0, 1.0 / dSqr);
+                    next.x = (1.0 - rInv)*next.x + rInv*curr.x;
+                    next.y = (1.0 - rInv)*next.y + rInv*curr.y;
+                    next.z = (1.0 - rInv)*next.z + rInv*curr.z;
+                }
+            }
+            delta = distance3d(curr, next);
+            curr.x = next.x;
+            curr.y = next.y;
+            curr.z = next.z;
+        }
+        return delta;
+    }
+
+    private static Coordinate initGuess(Coordinate[] points) {
+        Coordinate guess = new Coordinate(0, 0, 0);
+        for (Coordinate point : points) {
+            guess.x += point.x / points.length;
+            guess.y += point.y / points.length;
+            guess.z += point.z / points.length;
+        }
+        return guess;
+    }
+
+    private static Coordinate[] extractCoordinates(Geometry geometry) {
+        Coordinate[] points = geometry.getCoordinates();
+        if(points.length == 0)
+            return points;
+        boolean is3d = !Double.isNaN(points[0].z);
+        Coordinate[] coordinates = new Coordinate[points.length];
+        for(int i = 0; i < points.length; i++) {
+            coordinates[i] = points[i].copy();
+            if(!is3d)
+                coordinates[i].z = 0.0;
+        }
+        return coordinates;
+    }
+
+    public static Point geometricMedian(Geometry geometry, double tolerance, int maxIter, boolean failIfNotConverged) throws Exception {
```

Comment:
If you want, you can add more comprehensive tests in sedona-common to check the correctness of this function.